druid/src/text/format.rs

```diff
@@ -0,0 +1,208 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Creating, interpreting, and validating textual representations of values.
+
+use std::str::FromStr;
+use std::sync::Arc;
+
+use super::Selection;
+use crate::Data;
+
+/// A trait for types that create, interpret, and validate textual representations
+/// of values.
+///
+/// A formatter has two responsiblities: converting a value into an appropriate
+/// string representation, and attempting to convert a string back into the
+/// appropriate value.
+///
+/// In addition, a formatter performs validation on *partial* strings; that is,
+/// it determines whether or not a string represents a potentially valid value,
+/// even if it is not currently valid.
+pub trait Formatter<T> {
+    /// Return the string representation of this value.
+    fn format(&self, value: &T) -> String;
+    /// Return the string representation of this value, to be used during editing.
+    ///
+    /// This can be used if you want the text to differ based on whether or not
+    /// it is being edited; for instance you might display a dollar sign when
+    /// not editing, but not display it during editing.
+    fn format_for_editing(&self, value: &T) -> String {
+        self.format(value)
+    }
+
+    /// Determine whether the newly edited text is valid for this value type.
+    ///
+    /// This always returns a [`Validation`] object which indicates if
+    /// validation was successful or not, and which can also optionally,
+    /// regardless of success or failure, include new text and selection values
+    /// that should replace the current ones.
+    ///
+    ///
+    /// # Replacing the text or selection during validation
+    ///
+    /// Your `Formatter` may wish to change the current text or selection during
+    /// editing for a number of reasons. For instance if validation fails, you
+    /// may wish to allow editing to continue, but select the invalid region;
+    /// alternatively you may consider input valid but want to transform it,
+    /// such as by changing case or inserting spaces.
+    ///
+    /// If you do *not* explicitly set replacement text, and validation is not
+    /// successful, the edit will be ignored.
+    ///
+    /// [`Validation`]: Validation
+    fn validate_partial_input(&self, input: &str, sel: &Selection) -> Validation;
+
+    /// The value represented by the input, or an error if the input is invalid.
+    ///
+    /// This must return `Ok()` for any string created by [`format`].
+    ///
+    /// [`format`]: #tymethod.format
+    fn value(&self, input: &str) -> Result<T, ValidationError>;
+}
+
+/// The result of a [`Formatter`] attempting to validate some partial input.
+///
+/// [`Formatter`]: Formatter
+pub struct Validation {
+    result: Result<(), ValidationError>,
```

Comment:
This whole `Validation` struct is sort of an augmented result; it includes not only the `Result` itself, but also additional information that exists both in the success and failure cases. It would definitely be possible to store this as an `Option`, but I think keeping it as a `Result` more clearly reflects the semantics of the program.